File: 1.array/448.FindAllNumbersDisappearedInAnArray/main.py
from typing import List

# https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/

# TC: O(n), SC: O(1)
class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        n = len(nums) + 1
        for i, num in enumerate(nums):
            idx = abs(num) - 1
            if nums[idx] > 0:
                nums[idx] = -nums[idx]
        
        res = []
        for i in range(1, n):
            if nums[i-1] > 0:
                res.append(i)
        return res

# Cyclic sort: swap nums[i], nums[nums[i]-1] until nums[i] = nums[nums[i]-1]
# TC: O(n), SC: O(n)
class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        n = len(nums)
        for i in range(n):
            while nums[i] != nums[nums[i]-1]:
                # wrong:  nums[i], nums[nums[i]-1] = nums[nums[i]-1], nums[i]
                nums[nums[i]-1], nums[i]  = nums[i], nums[nums[i]-1]

        return [i+1 for i in range(n) if nums[i] != i+1]

File: 1.array/448.FindAllNumbersDisappearedInAnArray/main.go
package main

import "fmt"

// https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/description/

// [4,3,2,7,8,2,3,1] => [5,6]

// we will convert the position of the number value into negative number to mark it was appeared in array

// [4,3,-2,-7,8,2,3,1]
// i = 0; nums[nums[i-1]] = negative of nums[i-1]
// Finally, traverse from i: 1 -> n, nums[i-1] > 0 => push to the result array

// TC: O(n), SC: O(1) exclude variable of result

func findDisappearedNumbers(nums []int) []int {
	for _, num := range nums {
		index := num - 1
		if num < 0 {
			index = -num - 1
		}
		if nums[index] > 0 {
			nums[index] = -nums[index]
		}
	}

	var result []int
	for i := 1; i <= len(nums); i++ {
		if nums[i-1] > 0 {
			result = append(result, i)
		}
	}

	return result
}

func main() {
	fmt.Println(findDisappearedNumbers([]int{4, 3, 2, 7, 8, 2, 3, 1}))
}
